### PR TITLE
Improved cleaning of tombstones

### PR DESCRIPTION
**Pull request by abdelr**
### What's being changed:

Currently, we iterate over all nodes to find those connected to the removed nodes. Additionally, we search for the closest neighbours by performing a complete search on the graph.

This PR introduces two main changes. First, it rebuilds the section of the graph by only inferring the connections in the neighbourhood of the removed nodes based on the information available in the graph, rather than performing a search. 

Second, it runs the repair of the nodes concurrently since no search is need any longer.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [x] Chaos pipeline run or not necessary. Link to pipeline: https://github.com/weaviate/weaviate-chaos-engineering/actions/runs/7900483604
- [x] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

* relates to #3947

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/neighbor_connections.go [hunk 1]
```diff
@@ -13,20 +13,32 @@ package hnsw
...
 	denyList helpers.AllowList,
 ) error {
 	nfc := newNeighborFinderConnector(h, node, entryPointID, nodeVec, distancer, targetLevel,
-		currentMaxLevel, denyList)
+		currentMaxLevel, denyList, false)
+
+	return nfc.Do()
+}
+
+func (h *hnsw) findAndConnectNeighborsAfterCleanUpTombstonedNodes(node *vertex,
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/neighbor_connections.go:36]: Are we able to use something a shorter name for this? What about `findAndConnectNeighborsAfterTombstoneCleanup`?

File: adapters/repos/db/vector/hnsw/neighbor_connections.go [hunk 2]
```diff
@@ -42,21 +54,23 @@ type neighborFinderConnector struct {
...
-		denyList:        denyList,
+		graph:                       graph,
+		node:                        node,
+		entryPointID:                entryPointID,
+		nodeVec:                     nodeVec,
+		distancer:                   distancer,
+		targetLevel:                 targetLevel,
+		currentMaxLevel:             currentMaxLevel,
+		denyList:                    denyList,
+		afterCleanUpTombstonedNodes: afterCleanUpTombstonedNodes,
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/neighbor_connections.go:73]: Similar to above does `tombstoneCleanupNodes` work better for a name?

File: adapters/repos/db/vector/hnsw/search.go [hunk 3]
```diff
@@ -441,14 +442,7 @@ func (h *hnsw) distanceToFloatNode(distancer distancer.Distancer,
 ) (float32, bool, error) {
 	candidateVec, err := h.vectorForID(context.Background(), nodeID)
 	if err != nil {
-		var e storobj.ErrNotFound
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/search.go:444]: Are we fine deleting this section? Don't we still have to call `handleDeletedNode` when the node is not in the cache?

File: adapters/repos/db/vector/hnsw/delete.go [hunk 4]
```diff
@@ -293,22 +293,27 @@ func (h *hnsw) reassignNeighborsOf(deleteList helpers.AllowList, breakCleanUpTom
...
 	h.RUnlock()
+	h.resetLock.Lock()
+	defer h.resetLock.Unlock()
 
-	for n := 0; n < size; n++ {
-		if ok, err := h.reassignNeighbor(uint64(n), deleteList, breakCleanUpTombstonedNodes); err != nil {
-			return false, errors.Wrap(err, "reassign neighbor edges")
-		} else if !ok {
-			return false, nil
+	compressionhelpers.Concurrently(uint64(size), func(deletedID uint64) {
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/delete.go:299]: Should we add an environment variable to limit concurrency here?
Is 25% a better default than 100% total cores?

File: adapters/repos/db/vector/hnsw/search.go (hunk 3 above)
Comment (abdelr) [reply adapters/repos/db/vector/hnsw/search.go:444]: We do, but we handle the error where we have the information to take action. Here we are just hiding the reason of the error and the caller cannot handle it properly.

File: adapters/repos/db/vector/hnsw/neighbor_connections.go (hunk 1 above)
Comment (abdelr) [reply adapters/repos/db/vector/hnsw/neighbor_connections.go:36]: What about `reconnectNeighboursOf`. I have also simplified unused parameters.

**Review by trengrj — APPROVED** [review]

**Pull request merged**